### ISSUE TITLE
Seamless initialization from multiple places

### ISSUE THREAD

**Issue by ilearnio** [ilearnio/module-alias#35]
Allow module-alias to be initialized from multiple places, including separate packages in node_modules and nested/child packages (within source folder). It should add aliases to some global stack that will be shared between module-alias initializations. That will solve a lot of existing problems.

Related #32

**Comment by dorian-marchal** [comment]
Where should I start if I want to try to fix this issue, @Kehrlann?

**Comment by Kehrlann** [comment]
Hey @dorian-marchal, thanks for contributing. I've been off for a while, but I'll be back end of next week, I'll try and reply then.

**Comment by Kehrlann** [comment]
Hi @dorian-marchal !

Ok I just took a look at this. To be fair I'm not familiar enough with the ins and outs of the codebase to easily guide you through this issue. In particular, I'm not familiar with configuring module-alias programmatically, I've mostly worked on the "init" part of the codebase. So there is a little bit of research needed to better understand what's happening exactly.

The first step I would take would be to create a little github repo that showcases how module-alias fails when used in libraries/dependencies; and do a write up of the reasons behind the failures. If they're easy to replicate in the current test suite, then that's a next step.

If you don't feel comfortable doing the research bit yourself, I could look into it myself and explain my understanding here, but I have low bandwidth at the moment so I can't promise anything.

Let me know !

**Comment by Yehonal** [comment]
Any news about it?

**Comment by Kehrlann** [comment]
So I've taken a look at this. 

tl;dr: module-alias already works with library code

Here are my learnings:
1. It's not possible to isolate different "instances" of module-alias. Mainly because you don't know _WHICH MODULE_ is calling you. Even if you had a dict per module using module-alias, you wouldn't be able to determine which registry to use when "require" is called. The only way to do so would be to use a stack trace, and that's very hacky and fragile.
2. However, it is possible to use module-alias within a library. There are a few caveats though:
  1. As the aliases are global, you should probably prefix your aliases to avoid conflicts. Conflicts with end-user code, or with other libraries using module-alias. For example, you could prefix your aliases with '@my-lib/', and then use require('@my-lib/deep')
  2. The "register" mechanism does not work, as it tries to find package.json in either the current working directory of your node process, or two levels down from node_modules/module-alias. It is extremely likely that this is end-user code. So, instead, your should either register aliases manually with `moduleAlias.addAlias`, or using something like `require('module-alias')(__dirname).

I'll update the docs accordingly, and maybe contribute a sample project.

Cheers !

**Comment by ilearnio** [comment]
@Kehrlann Well, technically it should still be possible to define aliases as you described. However the current implementation is really not good for the job, my concerns are in possible conflicts and unpredictable behavior.

I guess I should have described my ideas of implementation more detailed. My original idea of how to make it work in other NPM packages was to use `global` object (something like `global.__moduleAlias = { ... }`) as a registry of all aliases, because there could be cases when different versions of module-alias are used at the same time and in the same project. I mean the following scenario:

```
node_modules/module-alias                                      (v2.2.2)
node_modules/child1/node_modules/module-alias                  (v2.0.1)
node_modules/child2/node_modules/module-alias                  (v2.0.0)
```  

In this case, when 3 different versions of module-alias are used in code, then 3 different `module-alias/index.js` files will be executed which will result in having 3 different registries of aliases and `Module._resolveFilename` being overridden 3 times. Technically, the aliases would still work, and each module-alias package will have it's own encapsulated registry, but it may result in unpredictable behavior since all 3 module-alias packages will modify `Module._resolveFilename`. And now imagine that they all will check for aliases in calls to `require`/`import`. Very unpredictable.

Here is one potential pitfall, if at some point the two children packages will eventually update module-alias to same latest version as main project has, then all aliases will be no longer be in 3 different registries but in 1 common registry, so their order may change.

The solution could be:

1. Create global registry for aliases that will be reused across all module-alias versions.
2. Allow to initialize (i.e. calling `init` method) the package multiple times. It can be called from the same package more than once or from other versions.
3. Do not modify `Module._nodeModulePaths` or `Module. _resolveFilename ` if they were already modified by another module-alias version. They will be modified once and will use the same global registry.

Aside:

- Display some kind of warning when trying to register same alias more than once.
- This should probably go to next version 3.0, since it's a breaking change (as global registry will not be compatible with previous versions).

**Comment by Kehrlann** [comment]
I see. That's a bigger change indeed. Maybe I'll see if #51 lands soon, to work on top of that change.

**Comment by Kehrlann** [comment]
Reopening, #85 does not fix this when multiple versions of `module-alias` coexist.

**Comment by ilearnio** [comment]
Hm, I guess it was auto-closed once I merged #85
